contracts/token/ERC721/ERC721.sol

```diff
@@ -330,32 +296,53 @@ contract ERC721 is Context, ERC165, IERC721, IERC721Metadata {
+    function _transfer(address from, address to, uint256 tokenId) internal {
         require(ERC721.ownerOf(tokenId) == from, "ERC721: transfer from incorrect owner");
         require(to != address(0), "ERC721: transfer to the zero address");
 
-        _beforeTokenTransfer(from, to, tokenId, 1);
+        _update(from, to, tokenId, 1);
+    }
 
-        // Check that tokenId was not transferred by `_beforeTokenTransfer` hook
-        require(ERC721.ownerOf(tokenId) == from, "ERC721: transfer from incorrect owner");
+    /**
+     * @dev Transfers `tokenId` from `from` to `to`, or alternatively mints (or burns) if `from` (or `to`) is
+     * the zero address. All customizations to transfers, mints, and burns should be done by overriding this function.
+     *
+     * Emits a {Transfer} event.
+     */
+    function _update(address from, address to, uint256 tokenId, uint256 batchSize) internal virtual {
+        if (batchSize > 1) {
+            if (from != address(0)) {
+                _balances[from] -= batchSize;
+            }
+            if (to != address(0)) {
+                _balances[to] += batchSize;
+            }
+        } else {
+            if (from != address(0)) {
+                // Clear approvals from the previous owner
+                delete _tokenApprovals[tokenId];
 
-        // Clear approvals from the previous owner
-        delete _tokenApprovals[tokenId];
-
-        unchecked {
-            // `_balances[from]` cannot overflow for the same reason as described in `_burn`:
-            // `from`'s balance is the number of token held, which is at least one before the current
-            // transfer.
-            // `_balances[to]` could overflow in the conditions described in `_mint`. That would require
-            // all 2**256 token ids to be minted, which in practice is impossible.
-            _balances[from] -= 1;
-            _balances[to] += 1;
-        }
-        _owners[tokenId] = to;
+                unchecked {
+                    // Cannot overflow, as that would require more tokens to be burned/transferred
+                    // out than the owner initially received through minting and transferring in.
+                    _balances[from] -= 1;
+                }
+            }
+
+            if (to != address(0)) {
+                unchecked {
+                    // Will not overflow unless all 2**256 token ids are minted to the same owner.
+                    // Given that tokens are minted one by one, it is impossible in practice that
+                    // this ever happens. Might change if we allow batch minting.
+                    // The ERC fails to describe this case.
+                    _balances[to] += 1;
+                }
+            }
 
-        emit Transfer(from, to, tokenId);
+            _owners[tokenId] = to;
 
-        _afterTokenTransfer(from, to, tokenId, 1);
+            emit Transfer(from, to, tokenId);
+        }
     }
```

Comment:
I'd do

```
function _update(address from, address to, uint256 tokenId, uint256 batchSize) internal virtual {
    if (batchSize == 1) {
        if (from != address(0)) {
            // Clear approvals from the previous owner
            delete _tokenApprovals[tokenId];

            unchecked {
                // Cannot overflow, as that would require more tokens to be burned/transferred
                // out than the owner initially received through minting and transferring in.
                _balances[from] -= 1;
            }
        }

        if (to != address(0)) {
            unchecked {
                // Will not overflow unless all 2**256 token ids are minted to the same owner.
                // Given that tokens are minted one by one, it is impossible in practice that
                // this ever happens. Might change if we allow batch minting.
                // The ERC fails to describe this case.
                _balances[to] += 1;
            }
        }

        _owners[tokenId] = to;

        emit Transfer(from, to, tokenId);
    } else if (batchSize > 1) {
        if (from != address(0)) {
            _balances[from] -= batchSize;
        }
        if (to != address(0)) {
            _balances[to] += batchSize;
        }
    }
    // else case (batchSize == 0) is a no-op, so we do nothing.
}
```

Reply:
I would say let's go with the first option.